### ISSUE TITLE
May 12 Source SDK Base 2013 Update seems to have broken binary compatibility

### ISSUE THREAD

**Issue by TomyLobo** [ValveSoftware/source-sdk-2013#374]
http://store.steampowered.com/news/21827/

This update seems to have contained more changes than advertised.
For me, a previously working game now crashes right after initializing the lightmaps.

Here's a diff between the fgd files before and after the update:
````diff
diff --git a/bin/base.fgd b/bin/base.fgd
index 19e7b46..47c1dcf 100755
--- a/bin/base.fgd
+++ b/bin/base.fgd
@@ -810,6 +810,9 @@
 	output OnStartTouchAll(void) : "Fired when an entity starts touching this trigger, and no other entities are touching it. Only entities that passed this trigger's filters are considered."
 	output OnEndTouch(void) : "Fired when an entity stops touching this trigger. Only entities that passed this trigger's filters will cause this output to fire."
 	output OnEndTouchAll(void) : "Fires when an entity stops touching this trigger, and no other entities are touching it. Only entities that passed this trigger's filters are considered."
+
+	// Inputs
+	input DisableAndEndTouch(void) : "Disables this trigger and calls EndTouch on all currently-touching entities."
 ]
 
 @BaseClass = worldbase
@@ -2443,6 +2446,12 @@
 		2 :  "No Dynamic Light" : 1
 	]
 
+	IgnoreSolid(choices) : "Ignore Solid" : 0 : "If set, this spotlight won't trace for solids." =
+	[
+		0 : "No"
+		1 : "Yes"
+	]
+
 	spotlightlength(integer) : "Spotlight Length" : 500 : "Length of the spotlight beam."
 	spotlightwidth(integer) : "Spotlight Width" : 50 : "Width of the spotlight beam."
 	rendercolor(color255) : "Color (R G B)" : "255 255 255"
@@ -5317,6 +5326,13 @@
 		0 : "No"
 		1 : "Yes"
 	]
+	generatelightmaps(choices) : "Generate (and use) lightmaps for this static prop" : 0 =
+	[
+		0 : "No"
+		1 : "Yes"
+	]
+	lightmapresolutionx(integer) : "Lightmap Resolution X" : 32 : "The resolution of the generated lightmap in the X (or U) direction (only used if Generate Lightmaps is Yes) "
+	lightmapresolutiony(integer) : "Lightmap Resolution Y" : 32 : "The resolution of the generated lightmap in the Y (or V) direction (only used if Generate Lightmaps is Yes) "
 ]
 
 @BaseClass = BaseFadeProp
@@ -6219,6 +6235,13 @@
 	]
 ]
 
+@SolidClass base(Trigger) = trigger_teleport_relative : 
+	"A trigger volume that teleports entities that touch it. Entities are teleported by Offset from their current position. Useful to get " +
+	"objects out of situations where levels or entities suddenly appear interseting with them."
+[
+	teleportoffset(vector) : "Offset (X Y Z)" : "0 0 0" : "The offset (in World-Space Coordinates X Y Z) to teleport the entity when the trigger is touched."
+]
+
 @SolidClass base(Targetname) = trigger_transition : 
 	"A volume that's used to control which entities go through the level transition. Create one or more trigger_transitions and "+
 	"give them the same name as the changelevel landmark. Any entities within the trigger_transition(s) will go to the next map."+
@@ -6544,3 +6567,12 @@
 	parm9(instance_parm) : "Parm" : : "This is a parameter.  It goes in the form of $variable type."
 	parm10(instance_parm) : "Parm" : : "This is a parameter.  It goes in the form of $variable type."
 ]
+
+@SolidClass base(Trigger) = trigger_apply_impulse : 
+	"A trigger that adds impulse to touching entities."
+[
+	impulse_dir(angle) : "Impulse Direction (Pitch Yaw Roll)" : "0 0 0" : "Angles indicating the direction to push touching entities."
+	force(float) : "Impulse Force" : "300" : ""
+
+	input ApplyImpulse(void)  : "Apply impulse to touching entities"
+]
````

I'm fairly new to source engine modding, but I think I see a pattern:
I get a crash right after a console message about light map initialization and the fgd file lists changes indicating something has been done with the lightmaps.

A mapper on the team says CS:GO has a feature called "static prop lightmaps", which previous titles lacked, so maybe that slipped into the release without making it into the update announcement.

tl;dr Whatever the May 12 update did, it broke binary compatibility. Binaries that worked with the previous version no longer work with this version.

**Comment by geotavros** [comment]
"static prop lightmaps" feature have been added before this update. But I don't remember was it in beta branch or release

**Comment by TomyLobo** [comment]
I can only tell that it wasn't in the fgd before.
What I compared was the "previous" branch and the default branch.
> The **previous SDK 2013 release branch** as well as the outdated "upcoming" branch are now available for a short time as "**previous**" and "previous_upcoming" respectively.

**Comment by Galaco** [comment]
This has broken counterstrike: source mapping. Adding new prop_statics/modifying an existing one (such that it adds missing key/values?) in a map, then compiling will crash the map on loading it. Compiling a map that has no changes since before the update will run fine.

I guess the short term solution is to disable lightmaps on every static_prop in the map (my primary project has >3500).

Would be great if someone else could confirm this to be true?

**Comment by StrikerMan780** [comment]
Stuff in the FGD itself isn't the issue, since those files aren't used by the engine, just the level editor (Hammer). However, it may be possible that new features in the engine itself may have broken things.

**Comment by TomyLobo** [comment]
> Stuff in the FGD itself isn't the issue, since those files aren't used by the engine, just the level editor (Hammer).

Obviously.

> However, it may be possible that new features in the engine itself may have broken things. (Which those new entries correspond to).

Exactly.

> However, good luck in ever getting Valve to fix anything.

Well I hope I get at least a reply explaining why they won't fix it.

**Comment by geotavros** [comment]
simply recompiling your mod and your maps doesn't fix your crash?

**Comment by TomyLobo** [comment]
That's not the point. The point is it's not backwards-compatible to begin with.

Qt programs don't crash with new Qt versions either, so why do source mods have to crash with new source versions?

Every time the SDK updates, we lose a generation of mods.
I think that's horrible.

**Comment by StrikerMan780** [comment]
That's strange, because I'm still able to run most older mods. Also, my own Source Engine mod, after the new update, still runs, even though it was made for the previous update.

**Comment by dotn3b** [comment]
The old release (now in the "**previous**" _beta_ branch) had some serious bugs that were being exploited to upload and execute code on multiplayer sessions' clients' machines. This has been known for a long while now and is in part - perhaps even mainly - what prompted the "**upcoming**", "**beta_test**" and "**prerelease**" beta updates.

The jumps to "upcoming" and "beta_test" were compatibility breaking which is why they were put into beta branches first and also why we got code updates via this repository to update our mods.
After two and a half years (**!**) that old exploitable version has been finally retired from the release branch and replace with the latest version. Again the old version is allegedly still available in form of the "previous" beta for players to use.

**In short:** This is not a bug, it's on purpose. Old mods that still rely on the _2013 Multiplayer Base_ should now be able to run using the _**previous**_ beta of the SDK **or preferably be updated** by the mod team if time allows and at all possible.

Certainly this is less than ideal but honestly a lot has been with the Source SDK for years now IMHO. Could you at least verify for us here that your mod works when switching to the "previous" branch? If it doesn't, I'm sure that's something someone at Valve (Neph?) will gladly fix.

**Comment by TomyLobo** [comment]
The "previous" branch works.

**Comment by Nephyrin** [comment]
Ideally, new SDK updates would retain binary compatibility with older versions, but the reality of how the Source engine works means this is not always the case.

If you have an example of a mod that is now crashing, you should open an issue with crash reports here so the cause of the incompatibility can be investigated.

**Comment by TomyLobo** [comment]
[Double Action: Boogaloo] (store.steampowered.com/app/317360), current version, comes with its own copy of source sdk, which i think is equivalent to the "previous" branch of the source sdk base, with which it also works, crashes with the current version of the source sdk base.

**Comment by StrikerMan780** [comment]
If only Valve made it so mods can specify what version of the SDK Base they need in the gameinfo.txt file, and load the correct binaries (without needing to swap branches)... that could prevent stuff like this from happening.

**Comment by StrikerMan780** [comment]
*sigh* There's no wonder why Valve gets an F- with the Better Business Bureau... their support for everything is complete and total ass.
